craft_parts/overlays/overlay_manager.py

```diff
@@ -0,0 +1,114 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Overlay mount operations and package installation helpers."""
+
+import logging
+from pathlib import Path
+from typing import List, Optional
+
+from craft_parts.infos import ProjectInfo
+from craft_parts.parts import Part
+
+from .overlay_fs import OverlayFS
+
+logger = logging.getLogger(__name__)
+
+
+class OverlayManager:
+    """Execution time overlay mounting and package installation.
+
+    :param project_info: The project information.
+    :param part_list: A list of all parts in the project.
+    :param base_layer_dir: The directory containing the overlay base, or None
+        if the project doesn't use overlay parameters.
+    """
+
+    def __init__(
+        self,
+        *,
+        project_info: ProjectInfo,
+        part_list: List[Part],
+        base_layer_dir: Optional[Path]
```

Comment:
why is this optional?

Reply:
The overlay manager is currently instantiated on the `Executor` class and passed to each part handler, even if the base layer directory is not specified by the user. This can probably be refactored to make the entire overlay manager instance to be optional in this case, but let's do that in a final refactoring round after all the overlay subsystem pieces are in place so we can better evaluate it in context.